src/core/Provider.ts

```diff
@@ -1,51 +1,87 @@
 
-const getState = (state: State) => state
-// TODO this doesn't trigger re-render so useDebugState can be stale.
-const subscribe = (_state: State, _callback: () => void) => () => {}
-
 export const Provider: React.FC<{
   initialValues?: Iterable<readonly [AnyAtom, unknown]>
   scope?: Scope
 }> = ({ initialValues, scope, children }) => {
   const storeRef = useRef<ReturnType<typeof createStore> | null>(null)
-  if (storeRef.current == null) {
-    // lazy initialization
-    storeRef.current = createStore(initialValues)
-  }
-  const store = storeRef.current as ReturnType<typeof createStore>
 
-  if (typeof process === 'object' && process.env.NODE_ENV !== 'production') {
-    // eslint-disable-next-line react-hooks/rules-of-hooks
-    useDebugState(useMutableSource(store[0], getState, subscribe))
+  if (
+    typeof process === 'object' &&
+    process.env.NODE_ENV !== 'production' &&
+    process.env.NODE_ENV !== 'test'
+  ) {
+    /* eslint-disable react-hooks/rules-of-hooks */
+    const atomsRef = useRef<AnyAtom[]>([])
+    if (storeRef.current == null) {
+      // lazy initialization
+      storeRef.current = createStore(initialValues, (newAtom) => {
+        atomsRef.current.push(newAtom)
+      })
+    }
+    useDebugState(
+      storeRef.current as ReturnType<typeof createStore>,
+      atomsRef.current
+    )
+    /* eslint-enable react-hooks/rules-of-hooks */
+  } else {
+    if (storeRef.current == null) {
+      // lazy initialization
+      storeRef.current = createStore(initialValues)
+    }
   }
 
   const StoreContext = getStoreContext(scope)
-  return createElement(StoreContext.Provider, { value: store }, children)
+  return createElement(
+    StoreContext.Provider,
+    { value: storeRef.current as ReturnType<typeof createStore> },
+    children
+  )
 }
 
 const atomToPrintable = (atom: AnyAtom) => atom.debugLabel || atom.toString()
 
-const stateToPrintable = (state: State) =>
+const stateToPrintable = ([state, atoms]: [State, AnyAtom[]]) =>
   Object.fromEntries(
-    Array.from(state.m.entries()).map(([atom, mounted]) => {
+    atoms.flatMap((atom) => {
+      const mounted = state.m.get(atom)
+      if (!mounted) {
+        return []
+      }
       const dependents = mounted.d
       const atomState = state.a.get(atom) || ({} as AtomState)
       return [
-        atomToPrintable(atom),
-        {
-          value: atomState.e || atomState.p || atomState.w || atomState.v,
-          dependents: Array.from(dependents).map(atomToPrintable),
-        },
+        [
+          atomToPrintable(atom),
+          {
+            value: atomState.e || atomState.p || atomState.w || atomState.v,
+            dependents: Array.from(dependents).map(atomToPrintable),
+          },
+        ],
       ]
     })
   )
 
-const useDebugState = (state: State) => {
-  useDebugValue(state, stateToPrintable)
+const getState = (state: State) => ({ ...state }) // shallow copy
+
+const useDebugState = (store: Store, atoms: AnyAtom[]) => {
```

Comment:
Maybe add a comment about that this makes sure we keep a reference to the atom in devmode, so it isn't GC'ed by the WeakMap

Reply:
great idea. is it allowed for you to add a suggestion?

Reply:
```suggestion
// This hook makes sure that we keep a reference to the atoms in dev mode, 
// so atoms aren't garbage collected by the WeakMap of mounted atoms
const useDebugState = (store: Store, atoms: AnyAtom[]) => {
```